### DISCUSSION TITLE
best way to introduce channels

### DISCUSSION THREAD

**Discussion by ditschedev** [uNetworking/uWebSockets.js#574] [Q&A] [answered]
As I read that topics are costly. 

Whats better to implement a „channel“ approach like pusher (e.g. a channel named ‘users/1/notifications’ or ‘tickets/345/watch’)? 

1. Using a topic per channel and publish to that topic.
2. Creating a Map with the key being the channel name and the value a Set of sockets. Publish via loop over sockets and `ws.send`.

I am currently using approach 1.

**Comment by e3dio** [comment, marked as answer]
Publish is best because iterating Sets can be slow, publish is most convenient and fastest, it will group messages by socket for you

**Reply by e3dio** [reply]
see @alexhultman did [benchmark](https://github.com/uNetworking/uWebSockets.js/discussions/553#discussioncomment-953576) and 100k topics for 100k users was almost the same speed as 1 topic for 100k users, therefor no reason to not use Publish from what I understand
